File: BackEnd/src/main/java/com/example/backend/services/SessionService.java
package com.example.backend.services;

import com.example.backend.dtos.ClassroomUpdateSaveDTO;
import com.example.backend.dtos.SessionDTO;
import com.example.backend.dtos.SessionNodeDto;
import com.example.backend.dtos.SessionSaveUpdateDTO;
import com.example.backend.entites.Session;
import org.springframework.data.util.Pair;

import java.util.List;

public interface SessionService {
    void saveSession(SessionSaveUpdateDTO sessionSaveUpdateDto, Long challengeId, Long classroomId);
    void updateSession(SessionSaveUpdateDTO sessionSaveUpdateDto);
    List<Session> getAllSessions();
    Pair<List<SessionNodeDto>, List<SessionNodeDto>> splitSessionByClassroomId(Long classroomId);
    List<SessionNodeDto> treenizationSessionByClassroomId(Long classroomId);

    List<SessionDTO> getSessionsDTOByClassroomId(Long classroomId);
    SessionNodeDto getSessionById(Long sessionId);
    void deleteSession(Long sessionId);
}

File: projectAngular/src/app/services/session.service.ts
// src/app/services/session.service.ts

import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SessionDTO } from '../entities/Session';
import { SessionNodeDto } from '../entities/SessionNodeDto';
import { SessionSaveUpdateDTO } from '../entities/sessionSaveUpdateDTO';


@Injectable({
    providedIn: 'root'
})
export class SessionService {

    private apiUrl = `http://localhost:8080/api/sessions`;

    constructor(private http: HttpClient) { }

    // Get all sessions
    getAllSessions(): Observable<SessionDTO[]> {
        console.log(this.http.get<SessionDTO[]>(`${this.apiUrl}/all`))
        return this.http.get<SessionDTO[]>(`${this.apiUrl}/all`);
    }

    // Get a session by ID
    getSessionById(id: number): Observable<SessionDTO> {
        return this.http.get<SessionDTO>(`${this.apiUrl}/${id}`);
    }

    // Create a new session
    createSession(session: SessionSaveUpdateDTO, challengeId: number, classroomId: number): Observable<void> {
        const params = { challengeId: challengeId.toString(), classroomId: classroomId.toString() };
        return this.http.post<void>(this.apiUrl, session, { params });
    }

    // Delete a session by ID
    deleteSession(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    // Get static integer (for testing purposes)
    getStaticInteger(): Observable<number> {
        return this.http.get<number>(`${this.apiUrl}/int`);
    }



    // Get the treenization of sessions by classroom ID
    treenizationSessionByClassroomId(classroomId: number): Observable<SessionNodeDto[]> {
        return this.http.get<SessionNodeDto[]>(`${this.apiUrl}/treenization/${classroomId}`);
    }
}
